### PR TITLE
Fix locating files from the autoload/files key

### PR DESCRIPTION
**Pull request by MidnightDesign**

### CODE REVIEW

**Comment by MidnightDesign** [comment]
I just realized that my fix also repaired the classmaps. So I added a test for that as well.

File: src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php [hunk 1]
```diff
@@ -55,9 +55,9 @@ private function locateFilesInFilesInFilesDefinitions(array $locations) : Genera
         foreach ($locations as $location) {
             if (is_file($location)) {
                 yield $location;
+            } elseif (is_dir($location)) {
+                yield from $this->extractFilesFromDirectory($location);
             }
-
-            yield from $this->extractFilesFromDirectory($location);
```
Comment (Ocramius) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: What are other alternatives btw?

Comment (Ocramius) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: As in... It's not a file nor a directory: it's a.... ?

Comment (heiglandreas) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: ```is_link```

Comment (Ocramius) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: Which generally resolves to one of the two above, no?

Comment (heiglandreas) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: Nope: Testing ```is_file``` and ```is_dir``` check whether the given items are a regular file or a directory. 

```
$ touch foo
$ mkdir bar
$ ln -s /dev/urandom foobar
$ ls -al
total 8
drwxr-xr-x   5 heiglandreas  wheel  170 23 Jan 16:56 .
drwxrwxrwt  16 root          wheel  544 23 Jan 16:55 ..
drwxr-xr-x   2 heiglandreas  wheel   68 23 Jan 16:55 bar
-rw-r--r--   1 heiglandreas  wheel    0 23 Jan 16:55 foo
lrwxr-xr-x   1 heiglandreas  wheel   12 23 Jan 16:56 foobar -> /dev/urandom
$ pwd
/tmp/linktest
$ php -a
Interactive shell

php > var_Dump(is_dir('/tmp/linktest/foobar'));
bool(false)
php > var_Dump(is_file('/tmp/linktest/foobar'));
bool(false)
php > var_Dump(is_link('/tmp/linktest/foobar'));
bool(true)
```

Comment (MidnightDesign) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: `is_dir()` returns `true` for a link pointing to a directory and `false` for a link pointing to a file. For `is_file()` it's the other way around. So as expected. (In case that wasn't obvious.)

Comment (MidnightDesign) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: ...so I think there aren't any alternatives. I'd still leave the `elseif`in though, and not change it to 
```php
if (is_file($location)) {
    yield $location;
    continue;
}
yield from $this->extractFilesFromDirectory($location);
```
if you were implying that.

Comment (MidnightDesign) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: What did I do differently?
```
$ touch foo
$ ln -s foo bar
$ php -a
Interactive mode enabled

php > var_dump(is_file('bar'));
bool(true)
```

Comment (Ocramius) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: Now that I think of it... it may return `false` for both dirs and files if the symlink is broken.

Comment (MidnightDesign) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: Correct.

```
$ ln -s doesnotexist foo
$ php -a
Interactive mode enabled

php > var_dump(is_file('foo'));
bool(false)
php > var_dump(is_dir('foo'));
bool(false)
```

Comment (Ocramius) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: Dammit, them edge cases. Is it possible to test this abomination? Worth it at all?

Comment (MidnightDesign) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: You mean just adding a test, right? The code is fine right now in that case, IMO.

**Comment by Ocramius** [comment]
Yeap, just the test

File: src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php (hunk 1 above)
Comment (MidnightDesign) [inline src/ComposerRequireChecker/FileLocator/LocateComposerPackageSourceFiles.php]: I can add them if you want. (I wouldn't though : )

**Comment by Ocramius** [comment]
Then don't ;-) open-ended question

**Comment by MidnightDesign** [comment]
Are there any open issues or can this be merged?

**Comment by Ocramius** [comment]
LGTM 👍
